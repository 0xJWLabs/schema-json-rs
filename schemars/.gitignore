/target
**/*.rs.bk
Cargo.lock
/tests/*.actual.*