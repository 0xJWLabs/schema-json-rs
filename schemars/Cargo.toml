[package]
name = "schemars"
description = "Generate JSON Schemas from Rust code"
homepage = "https://graham.cool/schemars/"
repository = "https://github.com/GREsau/schemars"
version = "0.6.3"
authors = ["Graham Esau <gesau@hotmail.co.uk>"]
edition = "2018"
license = "MIT"
readme = "README.md"
keywords = ["rust", "json-schema", "serde"]
categories = ["encoding"]
build = "build.rs"

[dependencies]
schemars_derive = { version = "=0.6.3", path = "../schemars_derive" }
serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"

chrono = { version = "0.4", default-features = false, optional = true }
indexmap = { version = "1.2", optional = true }
either = { version = "1.3", optional = true }

[dev-dependencies]
pretty_assertions = "0.6.1"

[features]
impl_json_schema = []
# derive_json_schema will be removed in a later version
derive_json_schema = ["impl_json_schema"]

[[test]]
name = "chrono"
required-features = ["chrono"]

[[test]]
name = "indexmap"
required-features = ["indexmap"]

[[test]]
name = "either"
required-features = ["either"]

[[test]]
name = "schema_for_schema"
required-features = ["impl_json_schema"]

[package.metadata.docs.rs]
all-features = true
