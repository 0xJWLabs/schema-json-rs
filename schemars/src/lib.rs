pub type Map<K, V> = std::collections::BTreeMap<K, V>;
pub type Set<T> = std::collections::BTreeSet<T>;

mod flatten;
mod json_schema_impls;
#[macro_use]
mod macros;

/// JSON Schema generator and settings.
pub mod gen;
/// JSON Schema types.
pub mod schema;

pub use schemars_derive::*;

use schema::Schema;

/// A type which can be described as a JSON Schema document.
///
/// This is implemented for many Rust primitive and standard library types.
///
/// This can also be automatically derived on most custom types with `#[derive(JsonSchema)]`.
///
/// # Example
/// ```
/// use schemars::{schema_for, JsonSchema};
///
/// #[derive(JsonSchema)]
/// struct MyStruct {
///     foo: i32,
/// }
///
/// let my_schema = schema_for!(MyStruct);
/// ```
pub trait JsonSchema {
    /// Whether JSON Schemas generated for this type should be re-used where possible using the `$ref` keyword.
    ///
    /// For trivial types (such as primitives), this should return `false`. For more complex types, it should return `true`.
    /// For recursive types, this *must* return `true` to prevent infinite cycles when generating schemas.
    ///
    /// By default, this returns `true`.
    fn is_referenceable() -> bool {
        true
    }

    /// The name of the generated JSON Schema.
    ///
    /// This is used as the title for root schemas, and the key within the `definitions` property for subschemas.
    fn schema_name() -> String;

    /// Generates a JSON Schema for this type.
    ///
    /// If the returned schema depends on any [referenceable](JsonSchema::is_referenceable) schemas, then this method will
    /// add them to the [`SchemaGenerator`](gen::SchemaGenerator)'s schema definitions.
    ///
    /// This should not return a `$ref` schema.
    fn json_schema(gen: &mut gen::SchemaGenerator) -> Schema;

    /// Helper for generating schemas for flattened `Option` fields.
    ///
    /// This should not need to be called or implemented by code outside of `schemars`.
    #[doc(hidden)]
    fn json_schema_optional(gen: &mut gen::SchemaGenerator) -> Schema {
        Self::json_schema(gen)
    }
}

#[cfg(test)]
pub mod tests {
    use super::*;

    pub fn schema_object_for<T: JsonSchema>() -> schema::SchemaObject {
        schema_object(schema_for::<T>())
    }

    pub fn custom_schema_object_for<T: JsonSchema>(
        settings: gen::SchemaSettings,
    ) -> schema::SchemaObject {
        schema_object(custom_schema_for::<T>(settings))
    }

    pub fn schema_for<T: JsonSchema>() -> schema::Schema {
        custom_schema_for::<T>(Default::default())
    }

    pub fn custom_schema_for<T: JsonSchema>(settings: gen::SchemaSettings) -> schema::Schema {
        T::json_schema(&mut gen::SchemaGenerator::new(settings))
    }

    pub fn schema_object(schema: schema::Schema) -> schema::SchemaObject {
        match schema {
            schema::Schema::Object(o) => o,
            s => panic!("Schema was not an object: {:?}", s),
        }
    }
}
