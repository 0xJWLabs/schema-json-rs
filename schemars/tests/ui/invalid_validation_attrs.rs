use schemars::JsonSchema;

#[derive(JsonSchema)]
pub struct Struct1(#[validate(regex = 0, foo, length(min = 1, equal = 2, bar))] String);

#[derive(JsonSchema)]
pub struct Struct2(#[schemars(regex = 0, foo, length(min = 1, equal = 2, bar))] String);

fn main() {}
