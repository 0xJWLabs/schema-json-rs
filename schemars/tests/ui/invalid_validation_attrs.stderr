error: expected validate regex attribute to be a string: `regex = "..."`
 --> $DIR/invalid_validation_attrs.rs:4:39
  |
4 | pub struct Struct1(#[validate(regex = 0, foo, length(min = 1, equal = 2, bar))] String);
  |                                       ^

error: unknown schemars attribute `foo`
 --> $DIR/invalid_validation_attrs.rs:7:42
  |
7 | pub struct Struct2(#[schemars(regex = 0, foo, length(min = 1, equal = 2, bar))] String);
  |                                          ^^^

error: expected schemars regex attribute to be a string: `regex = "..."`
 --> $DIR/invalid_validation_attrs.rs:7:39
  |
7 | pub struct Struct2(#[schemars(regex = 0, foo, length(min = 1, equal = 2, bar))] String);
  |                                       ^

error: schemars attribute cannot contain both `equal` and `min`
 --> $DIR/invalid_validation_attrs.rs:7:63
  |
7 | pub struct Struct2(#[schemars(regex = 0, foo, length(min = 1, equal = 2, bar))] String);
  |                                                               ^^^^^^^^^

error: unknown item in schemars length attribute
 --> $DIR/invalid_validation_attrs.rs:7:74
  |
7 | pub struct Struct2(#[schemars(regex = 0, foo, length(min = 1, equal = 2, bar))] String);
  |                                                                          ^^^
